Updating the trigger documentation from v4 to v5 (#365)
DUNE-DAQ/trigger: README.md

> *(Last updated July 2023, MiR + AS)*

Reviewer: Should this be updated?

Author: Agreed & done. Added your name too.

> The `trigger` package contains the modules that make up the DUNE FD DAQ trigger system. Implementations of the physics algorithms that produce data selection objects (trigger primitive, trigger activity and trigger candidates) live in the [`tpglibs`](https://github.com/DUNE-DAQ/tpglibs) and [`triggeralgs`](https://github.com/DUNE-DAQ/triggeralgs) packages. The confoguration schema & code that generates trigger application live in [`appmodel`](https://github.com/DUNE-DAQ/appmodel)

Reviewer: ```suggestion
The `trigger` package contains the modules that make up the DUNE FD DAQ trigger system. Implementations of the physics algorithms that produce data selection objects (trigger primitive, trigger activity and trigger candidates) live in the [`tpglibs`](https://github.com/DUNE-DAQ/tpglibs) and [`triggeralgs`](https://github.com/DUNE-DAQ/triggeralgs) packages. The configuration schema & code that generates trigger application live in [`appmodel`](https://github.com/DUNE-DAQ/appmodel)
```

<hr>

> The main goal of the trigger is to extract information from data to form Trigger Decisions, requests to save the raw data. Additionally, in conjunction with the DataFilter, [`datafilter`](https://github.com/DUNE-DAQ/datafilter), the aim is to reduce the total data volume from the four planned DUNE modules to be in line with the DUNE requirements for trigger acceptance and efficiency for various physics signals.

Reviewer: ```suggestion
The main goal of the trigger is to extract information from data to form Trigger Decisions, which make the requests to save the raw data. Additionally, in conjunction with the DataFilter, [`datafilter`](https://github.com/DUNE-DAQ/datafilter), the aim is to reduce the total data volume from the four planned DUNE modules to be in line with the DUNE requirements for trigger acceptance and efficiency for various physics signals.
```

  <img src="docs/minimal.svg" />
</p>

> The hits-on-wire, in the form of Trigger Primitives, are merged at the scale of one TDAQ unit (such as an APA or CRP) into a cluster representing some type of Trigger Activity. Multiple Trigger Activity outputs can then be merged into a single Trigger Candidate that may include information spanning multiple TDAQ units or multiple slices of time (say, many APAs or many seconds) for a single system. Finally, multiple Trigger Candidates can be merged across systems (e.g. TPC and PDS) in the MLT into a Trigger Decision, a request to save the data.

Reviewer: Maybe change "hits-on-wire" to "hits-on-channel". The CRPs use copper strips instead of wires for their channels.

Author: Agreed & done. Found one more instance of using "wire" rather than "channel", so changed that too.

* **Trigger Record (TR)**: An object in a stored file, containing the raw data, TPs, TAs, and TCs that have led to its construction.

#### Modules-Makers:
* **Trigger Primitive Generator (TPG)**: An algorithm that sees the continous waveform on a channel, and detects "hits", generating Trigger Primitive (TP) object.

> * **Trigger Activity Maker (TAM)**: A module processing the incoming stream of TPs and finding a localised activity at the granularity of single APA/CRPs (and PDS detector units in the future). TAMs has associated algorithms, that define the search for activity (e.g. based on clustering). The TAs that the TAM produces are sent to the TCM.

Reviewer: ```suggestion
* **Trigger Activity Maker (TAM)**: A module processing the incoming stream of TPs and finding localised activity at the granularity of single APA/CRPs (and PDS detector units in the future). TAMs have associated algorithms that define the search for activity (e.g. based on clustering). The TAs that a TAM produces are sent to the TCM.
```

* **Module Level Trigger (MLT)**: A module that handles TCs and produces data readout signals (Trigger Decisions) based on those. This may include merging overlapping events, deciding on the readout windows, ROI (in the future) etc.

#### Modules-Other:
* **TriggerDataHandlerModule**: The trigger objects are contained inside independent "DataHandler" units. Each unit receives some type of data (e.g. TriggerPrimitive), can run pre-processing task on that data, places it into a buffer, and runs post-processing task after retreival from that buffer. The post-processing (e.g. TriggerActivityMaker) task usually generates different data-type, e.g. TriggerActivity. The TriggerDataHandlerModule also contains a LatencyBufferHandler, which deals with buffer cleanup, readout requests etc.

## Flow
There currently does not exist an up-to-date representation of the trigger system, but from the beginning of the v5 development:
<p align="center">
  <img src="docs/trigger.png" />
</p>

> The diagram above shows the **DAQModules** and connections in an example trigger and readout app. Blue rounded rectangles are the **TriggerDataHandler** **DAQModules**, red rectangle are the external trigger inputs, the orange rectangle represents the readout application, and the purple represents the **ModuleLevelTrigger**, handling the trigger candidates. Each **DataHandler** module (here called **ReadoutModel**) receives one type of data (e.g. TriggerActivity), runs pre-processing tasks on it, inserts it into a latency buffer (that is handled by the LatencyBufferHandler), and runs post-processing tasks generating a new object (e.g. TriggerCandidate).

Reviewer: ```suggestion
The diagram above shows the **DAQModules** and connections in an example trigger and readout app. Blue rounded rectangles are the **TriggerDataHandler** **DAQModules**, red rectangles are the external trigger inputs, the orange rectangle represents the readout application, and the purple represents the **ModuleLevelTrigger**, handling the trigger candidates. Each **DataHandler** module (here called **ReadoutModel**) receives one type of data (e.g. TriggerActivity), runs pre-processing tasks on it, inserts it into a latency buffer (that is handled by the LatencyBufferHandler), and runs post-processing tasks generating a new object (e.g. TriggerCandidate).
```

## Links to other repositories

### tpglibs ([link](https://github.com/DUNE-DAQ/tpglibs)) <a name="tpglibs"></a>

> Repository that holds the algorithms that generate the TriggerPrimitives out of the continously-supplied waveforms, from each wire/channel. It is used by the readout application, and includes both the naive and AVX implementations of the TPG algorithms.

Reviewer: ```suggestion
Repository that holds the algorithms that generate the TriggerPrimitives out of the continously-supplied waveforms, from each channel. It is used by the readout application, and includes both the native and AVX implementations of the TPG algorithms.
```

(I'll be trying to move away from the "naive" diction to "native".)

### triggeralgs ([link](https://github.com/DUNE-DAQ/triggeralgs)) <a name="triggeralgs"></a>

> The triggeralgs package contains implementations of the algorithms that produce TAs from TPs and TCs from TAs. They are in a separate package to allow the algorithms to be built outside of the DAQ software stack and used in the offline software, enabling a completely-independent development of the trigger algorithms. Trigger algorithms are easliely loaded dynamically only knowing their name, through a factory method: [AlgorithmPlugins.hpp](https://github.com/DUNE-DAQ/trigger/blob/develop/include/trigger/AlgorithmPlugins.hpp).

Reviewer: ```suggestion
The triggeralgs package contains implementations of the algorithms that produce TAs from TPs and TCs from TAs. They are in a separate package to allow the algorithms to be built outside of the DAQ software stack and used in the offline software, enabling a completely-independent development of the trigger algorithms. Trigger algorithms are easily loaded dynamically only knowing their name, through a factory method: [AlgorithmPlugins.hpp](https://github.com/DUNE-DAQ/trigger/blob/develop/include/trigger/AlgorithmPlugins.hpp).
```

### trgdataformats ([link](https://github.com/DUNE-DAQ/trgdataformats)) <a name="trgdataformats"></a>
The trigger data objects are defined as C++ structures, which could potentially be expressed in different ways in memory (eg on big-endian or little-endian systems), and do not have to be contiguous in memory.
The trgdataformats repository contains the C++ structures that hold the trigger data formats, and various overlays that extend the trigger objects to be continous in memory to allow them to be stored in the buffers and saved as fragments.

### appmodel ([link](https://github.com/DUNE-DAQ/appmodel)) <a name="appmodel"></a>

> The appmodel repository contains the configuration schema for all the trigger (an more) objects, together with functions that generate the trigger application out the xml data that follows those schemas. This includes configuration of the trigger algorithms, setup of the external triggers, creating queues between different `DAQModules` etc.

Reviewer: ```suggestion
The appmodel repository contains the configuration schema for all the trigger objects (and more), together with functions that generate the trigger application out of xml data that follows those schemas. This includes configuration of the trigger algorithms, setup of the external triggers, creating queues between different `DAQModules` etc.
```

* Daqconf wiki: [link](https://github.com/DUNE-DAQ/daqconf/wiki)

### offline
**Offline live emulation currently not supported in v5**

## Development
* [Development workflow](https://dune-daq-sw.readthedocs.io/en/latest/packages/daq-release/development_workflow_gitflow/)

> * [Setting up DUNE DAQ development area (v5.3)](https://github.com/DUNE-DAQ/daqconf/wiki/Setting-up-a-fddaq%E2%80%90v5.3.0-development-area)

Reviewer: This may be annoying to keep track of.

Author: Yeah. I've now changed the link to `https://github.com/DUNE-DAQ/daqconf/wiki`, so we hopefully won't have to keep track of it anymore.

* **Follow the styleguide** linked above.
* **Never push directly to the `develop` branch**. Create a branch (name should contain an easily recognisable name and reflect the purpose of the feature/fix, e.g. `name/new_feature`.) and make a **pull request** to the appropriate branch. At least one reviewer is required (more for big changes). General rule of thumb: don't merge your own PR.

> * **Always use integration tests.** A selection is available at [daq-systemtest integration tests](https://github.com/DUNE-DAQ/daqsystemtest/tree/develop/integtest). As a minimum, the `minimal_system_quick_test`, `fake_data_producer_test` and `3ru_3df_multirun_test` tests should be run (the more the better of course). Some tests require more powerful machines (available at np04).

Reviewer: ```suggestion
* **Always use integration tests.** A selection is available at [daq-systemtest integration tests](https://github.com/DUNE-DAQ/daqsystemtest/tree/develop/integtest). At a minimum, the `minimal_system_quick_test`, `fake_data_producer_test` and `3ru_3df_multirun_test` tests should be run (the more the better of course). Some tests require more powerful machines (available at np04).
```
